File: Data Structures & Algorithms/house-robber/submission-2.java
class Solution {
    public int rob(int[] nums) {
        int[] dp = new int[nums.length];
        Arrays.fill(dp, -1);
        return helper(0, nums, dp);
    }

    public int helper(int idx, int[] nums, int[] dp){
        if(idx >= nums.length)
            return 0;
        if(dp[idx] != -1)
            return dp[idx];
        dp[idx] = Math.max(
            nums[idx] + helper(idx + 2, nums, dp), 
            helper(idx + 1, nums, dp)
        );
        return dp[idx];
    }
}

File: Data Structures & Algorithms/target-sum/submission-2.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        memo = {}

        def helper(i, target):
            if i == len(nums):
                if target == 0:
                    return 1
                return 0
            if (i, target) in memo:
                return memo[(i, target)]
            memo[(i, target)] = helper(i + 1, target - nums[i]) + helper(i + 1, target + nums[i])
            return memo[(i, target)]

        return helper(0, target)